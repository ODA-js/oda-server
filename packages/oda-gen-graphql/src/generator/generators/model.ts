import { lib } from 'oda-gen-common';
import { Factory } from 'fte.js';
import * as template from '../../graphql-backend-template';
import * as path from 'path';
import { writeFile } from './writeFile';

const { get, deepMerge } = lib;

export default function $generateModel(
  raw,
  rootDir,
  model,
  typeMapper: { [key: string]: (q: string) => string },
  route: string,
  fileName: string,
) {
  let source = get(template, `model.${route}`).generate(raw, model, typeMapper);
  let fn = path.join(rootDir, fileName);

  if (typeof source === 'string') {
    fn = path.join(rootDir, fileName);
    writeFile(fn, source);
  } else if (Array.isArray(source)) {
    let parts = route.split('.').slice(1); // it is always `data`, at least here
    source.forEach(f => {
      fn = path.join(rootDir, f.name);
      writeFile(fn, f.content);
    });
  }
}
