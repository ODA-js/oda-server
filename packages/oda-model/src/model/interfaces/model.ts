import { MetaModelInput, ModelPackageInput, EntityBaseInput } from './input';
import { PackageMeta, EntityMeta, EntityBaseMeta, MixinMeta } from './metadata';

/**
 * the meta model of an application
 */
export interface IModel extends IModelBase<ModelMeta, MetaModelInput> {
  /**
   * set of packages within model
   */
  readonly packages: Map<string, IPackage>;
}
/**
 * Modeled package
 */
export interface IPackage extends IModelBase<PackageMeta, ModelPackageInput> {
  /**
   * is it is abstract
   */
  readonly abstract: boolean;
  /**
   * owner model
   */
  readonly metaModel: IModel;
  /**
   * set of entities
   */
  readonly entities: Map<string, IEntity>;
  /**
   * set of scalars
   */
  readonly scalars: Map<string, IScalar>;
  /**
   * set of mixins
   */
  readonly mixins: Map<string, IMixin>;
  /**
   * set of enums
   */
  readonly enums: Map<string, IEnum>;
  /**
   * set of unions
   */
  readonly unions: Map<string, IUnion>;
  /**
   * set of mutations
   */
  readonly mutations: Map<string, IMutation>;
  /**
   * available user queries
   */
  readonly queries: Map<string, IQuery>;
  /**
   * directives
   */
  readonly directives: Map<string, IDirective>;
}

/**
 * Enum definitions
 */

/**
 * enumItem definition
 */

/**
 * base for entity like items
 */
export interface IEntityBase<T extends EntityBaseMeta>
  extends IModelBase<EntityMeta, EntityBaseInput<T>> {
  /**
   * plural name for collection
   */
  plural: string;
  /**
   * title for pural
   */
  titlePlural: string;
  /**
   * set of fields
   */
  fields: Map<string, ISimpleField>;
}

/**
 * Interface definition
 */
/**
 * Entity definitions
 */
export interface IEntity extends IEntityBase<EntityMeta> {
  /**
   * set of interfaces/Mixin it is implement
   * so it is has no own methods to be received from storage
   */
  implements: Set<string>;
  /**
   * is this entity is only embedded into other entity
   */
  embedded: boolean;
  /**
   * entities it is embedded into
   */
  embeddedInto: Set<string>;
  /**
   * is this entity is abstract and can be used for descending from
   */
  abstract: boolean;
}

/**
 * the base for field-like items
 */

/**
 * entity field definition
 */

/**
 * entity reference
 */

// export type ModelItem =
//   | IModel
//   | IPackage
//   | IEntity
//   | IField
//   | IRelation<RelationMeta>;

// export type Relation =
//   | IHasManyRelation
//   | IHasOneRelation
//   | IBelongsToRelation
//   | IBelongsToRelation;

// export function isModel(item: ModelItem): item is IModel {
//   return item.modelType === 'model';
// }

// export function isPackage(item: ModelItem): item is IPackage {
//   return item.modelType === 'package';
// }

// export function isEntity(item: ModelItem): item is IEntity {
//   return item.modelType === 'entity';
// }

// export function isField(item: ModelItem): item is IField {
//   return item.modelType === 'field';
// }

// export function isRelation<T extends RelationMeta>(
//   item: ModelItem,
// ): item is IRelation<T> {
//   return (
//     item.modelType === 'BelongsTo' ||
//     item.modelType === 'BelongsToMany' ||
//     item.modelType === 'HasOne' ||
//     item.modelType === 'HasMany'
//   );
// }

// export function IsBelongsTo(item: Relation): item is IBelongsToRelation {
//   return isRelation(item) && item.modelType === 'BelongsTo';
// }

// export function IsBelongsToMany(item: Relation): item is IBelongsToRelation {
//   return isRelation(item) && item.modelType === 'BelongsToMany';
// }

// export function IsHasOne(item: Relation): item is IBelongsToRelation {
//   return isRelation(item) && item.modelType === 'HasOne';
// }

// export function IsHasMany(item: Relation): item is IBelongsToRelation {
//   return isRelation(item) && item.modelType === 'HasMany';
// }
