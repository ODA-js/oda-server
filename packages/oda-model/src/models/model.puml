@startuml
ModelBase ----|> Element
EntityReference --|> Element
RelationBase --|> Element
BelongTo --|> RelationBase
BelongToMany --|> RelationBase
HasOne --|> RelationBase
HasMany --|> RelationBase
Directive --|> ModelBase
Args --|> ModelBase
Enum ---|> ModelBase
Enum --* EnumItem
EnumItem --|> ModelBase
Mutation ---|> ModelBase
Operation ---|> ModelBase
PackageBase ----|> ModelBase
MetaModel --|>  PackageBase
ModelPackage --|> PackageBase
ModelHook --|> PackageBase
Query --|> ModelBase
Union --|> ModelBase
Scalar --|> ModelBase
EntityBase ----|> ModelBase
Entity --|> EntityBase
Mixin --|> EntityBase
FieldBase -----|> ModelBase
RelationFieldBase --|> FieldBase
SimpleField --|> FieldBase
RelationField --|> RelationFieldBase
EntityField --|> RelationFieldBase
InputType ----|> ModelBase
InputFIeld ----|> ModelBase

@enduml
