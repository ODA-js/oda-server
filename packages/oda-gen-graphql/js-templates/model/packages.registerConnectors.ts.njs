<#@ context 'model'#>
<#- for(let curr of model.packageList){#>
import #{curr.entry} from './data/registerConnectors';
<#- }#>

export default {
<#- for(let curr of model.packageList){#>
  #{curr.name}: #{curr.entry},
<#- }#>
};
