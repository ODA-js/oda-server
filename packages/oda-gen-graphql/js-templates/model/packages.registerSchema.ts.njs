<#@ context 'model'#>
<#- for(let curr of model.packageList){#>
import #{curr.entry} from './#{curr.name}/entity/registerSchema';
<#- }#>

export default {
<#- for(let curr of model.packageList){#>
  #{curr.name}: #{curr.entry},
<#- }#>
};
