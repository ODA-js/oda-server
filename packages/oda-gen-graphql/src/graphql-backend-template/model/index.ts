import * as registerConnectors from './packages.registerConnectors.ts';

export { registerConnectors };
