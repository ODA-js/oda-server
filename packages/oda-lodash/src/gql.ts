import { DocumentNode, Kind, parse, visit, getOperationAST } from 'graphql';

import { argumentsObjectFromField } from 'apollo-utilities';
import { each } from 'lodash';
import { get } from 'lodash';
import { isEqual } from 'lodash';
import { keyBy } from 'lodash';
import { set } from 'lodash';

import Lodash from './schema';
import { applyTransformations } from './transformations';
import { makeExecutableSchema } from 'graphql-tools';

let _lodashAST: any;

export function lodashAST() {
  if (!_lodashAST) {
    let current = Lodash;
    current.build();
    current.fixSchema();
    current.applyHooks();
    // fs.writeFileSync(`${pack}.graphql`, current.typeDefs.toString());
    const schema = makeExecutableSchema({
      typeDefs: current.schema,
      resolvers: current.resolvers,
      resolverValidationOptions: {
        requireResolversForNonScalar: false,
      },
    });
    _lodashAST = schema.getDirective('_');
  }
  return _lodashAST;
}

export function graphqlLodash(
  query: string | DocumentNode,
  operationName?: string,
) {
  const pathToArgs = {};
  const reshape = {};
  const queryAST = typeof query === 'string' ? parse(query) : query;
  traverseOperationFields(
    queryAST,
    operationName,
    (node: any, resultPath: any, alias: any) => {
      let args = getLodashDirectiveArgs(node, lodashAST());
      if (args === null) {
        return;
      }

      // TODO: error if transformation applied on field that already
      // seen without any transformation
      const argsSetPath = [...resultPath, '@_'];
      const previousArgsValue = get(pathToArgs, argsSetPath, null);
      if (previousArgsValue !== null && !isEqual(previousArgsValue, args)) {
        throw Error(
          `Different "@_" args for the "${resultPath.join('.')}" path`,
        );
      }
      set(pathToArgs, argsSetPath, args);
      set(reshape, [...resultPath, '_@'], alias[resultPath.length - 1]);
    },
  );
  return {
    apply: !!Object.keys(pathToArgs).length,
    query: stripQuery(queryAST),
    transform: (data: any) => applyLodashDirective(pathToArgs, data),
    queryShape: pathToArgs,
    reshape,
  };
}

function traverseOperationFields(queryAST: any, operationName: any, cb: any) {
  const fragments: any = {};
  const operationAST = getOperationAST(queryAST, operationName);
  if (!operationAST) {
    throw new Error(
      'Must provide operation name if query contains multiple operations.',
    );
  }

  queryAST.definitions.forEach((definition: any) => {
    if (definition.kind === Kind.FRAGMENT_DEFINITION) {
      fragments[definition.name.value] = definition;
    }
  });

  const resultPath: any[] = [];
  const alias: any[] = [];
  cb(operationAST, resultPath, alias);
  traverse(operationAST);

  function traverse(root: any) {
    visit(root, {
      enter(node) {
        if (node.kind === Kind.FIELD) {
          resultPath.push((node.alias || node.name).value);
          alias.push({ [(node.alias || node.name).value]: node.name.value });
        }

        if (node.kind === Kind.FRAGMENT_SPREAD) {
          const fragmentName = node.name.value;
          const fragment = fragments[fragmentName];
          if (!fragment) {
            throw Error(`Unknown fragment: ${fragmentName}`);
          }
          traverse(fragment);
        }
      },
      leave(node) {
        if (node.kind !== Kind.FIELD) {
          return;
        }
        cb(node, resultPath, alias);
        resultPath.pop();
        alias.pop();
      },
    });
  }
}

function getLodashDirectiveArgs(node: any, lodashDirectiveDef: any) {
  let lodashNode = null;

  for (let directive of node.directives) {
    if (directive.name.value !== lodashDirectiveDef.name) {
      continue;
    }
    if (lodashNode) {
      throw Error(`Duplicating "@_" on the "${node.name.value}"`);
    }
    lodashNode = directive;
  }

  if (lodashNode === null) {
    return null;
  }
  const args = argumentsObjectFromField(lodashNode, lodashDirectiveDef);
  return normalizeLodashArgs(lodashNode.arguments, args);
}

function normalizeLodashArgs(argNodes: any, args: any) {
  if (!argNodes) {
    return args;
  }

  //Restore order of arguments
  argNodes = keyBy(argNodes, argNode => argNode.name.value);
  const orderedArgs: any = {};
  each(argNodes, (node, name) => {
    const argValue = args[name];

    if (node.value.kind === 'ObjectValue') {
      orderedArgs[name] = normalizeLodashArgs(node.value.fields, argValue);
    } else if (node.value.kind === 'ListValue') {
      const nodeValues = node.value.values;
      orderedArgs[name] = [];
      for (let i = 0; i < nodeValues.length; ++i) {
        orderedArgs[name][i] = normalizeLodashArgs(
          nodeValues[i].fields,
          argValue[i],
        );
      }
    } else if (node.value.kind === 'EnumValue' && node.value.value === 'none') {
      orderedArgs[name] = undefined;
    } else {
      orderedArgs[name] = argValue;
    }
  });
  return orderedArgs;
}

function stripQuery(queryAST: any): DocumentNode {
  return visit(queryAST, {
    [Kind.DIRECTIVE]: (node: any) => {
      if (node.name.value === '_') {
        return null;
      }
      return;
    },
  });
}

function applyLodashDirective(pathToArgs: any, data: any) {
  if (data === null) {
    return null;
  }

  const changedData = applyOnPath(data, pathToArgs);
  return applyLodashArgs([], changedData, pathToArgs['@_']);
}

function applyOnPath(result: any, pathToArgs: any) {
  const currentPath: any[] = [];
  return traverse(result, pathToArgs);

  function traverse(root: any, pathRoot: any): any {
    // if (root === null || root === undefined) {
    //   return null;
    // }
    if (Array.isArray(root)) {
      return root.map(item => traverse(item, pathRoot));
    }

    if (typeof root === 'object') {
      const changedObject = root ? Object.assign({}, root) : root;
      for (const key in pathRoot) {
        if (key === '@_') {
          continue;
        }
        currentPath.push(key);

        let changedValue = traverse(get(root, key), get(pathRoot, key));
        // if (changedValue === null || changedValue === undefined){
        //   continue;
        // }

        const lodashArgs = get(pathRoot, [key, '@_'].join('.'));
        changedValue = applyLodashArgs(currentPath, changedValue, lodashArgs);
        if (typeof changedValue === 'function') {
          changedValue(changedObject, key);
        } else {
          set(changedObject, key, changedValue);
        }
        currentPath.pop();
      }
      return changedObject;
    } else {
      return root;
    }
  }
}

function applyLodashArgs(path: any, object: any, args: any) {
  try {
    return applyTransformations(object, args);
  } catch (e) {
    // FIXME:
    // tslint:disable-next-line:no-console
    console.log(path);
    throw e;
  }
}
