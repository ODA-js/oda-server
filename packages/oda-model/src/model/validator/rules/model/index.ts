import emptyName from './emptyName';
import defaultPackage from './defaultPackage';

export default [new emptyName(), new defaultPackage()];
