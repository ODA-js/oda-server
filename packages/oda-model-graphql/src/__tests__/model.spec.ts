import 'jest';

import gql from 'graphql-tag';
import model from '../schema';
import { reshape } from 'oda-lodash';

import { MetaModel } from 'oda-model';
const query = gql`
  {
    packages {
      name
    }
    entities @_(filter: { name: "MediaFiles" }) {
      metadata
      name
      props: fields {
        name
      }
    }
  }
`;

describe('model', () => {
  it('loads', () => {
    const repository = new MetaModel(model as any);
    // const resolver = (fieldName: string, root: any) => {
    //   return root[fieldName];
    // };

    const result = reshape(query, repository.toObject());
    expect(result).toMatchSnapshot('packages');
    expect(repository).toMatchSnapshot('repository');
    expect(repository.toObject()).toMatchSnapshot('repository.toObject');
  });
});
