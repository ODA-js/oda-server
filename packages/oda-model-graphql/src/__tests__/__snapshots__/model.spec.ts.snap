// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model loads: packages 1`] = `
Object {
  "entities": Array [
    Object {
      "metadata": Object {
        "UI": Object {
          "edit": Array [],
          "embedded": Array [],
          "hidden": Array [],
          "list": Array [],
          "listName": Array [
            "title",
          ],
          "quickSearch": Array [],
          "show": Array [],
        },
        "acl": Object {
          "create": Array [],
          "delete": Array [],
          "read": Array [],
          "upadte": Array [],
        },
        "name": Object {
          "plural": "MediaFiles",
          "singular": "MediaFiles",
        },
        "persistence": Object {
          "indexes": Object {
            "id": Object {
              "fields": Object {
                "id": 1,
              },
              "name": "id",
              "options": Object {
                "unique": true,
              },
            },
          },
        },
        "titlePlural": "MediaFiles",
      },
      "name": "MediaFiles",
      "props": Array [
        Object {
          "name": "title",
        },
        Object {
          "name": "description",
        },
        Object {
          "name": "reference",
        },
        Object {
          "name": "catalog",
        },
        Object {
          "name": "id",
        },
      ],
    },
  ],
  "packages": Array [],
}
`;
